into/backends/tests/test_aws.py

```diff
@@ -0,0 +1,253 @@
+
+
+@pytest.yield_fixture
+def tmpcsv():
+    with tmpfile('.csv') as fn:
+        with open(fn, mode='w') as f:
+            df.to_csv(f, index=False)
+        yield fn
+
+
+@pytest.fixture(scope='module')
+def conn():
+    # requires that you have a config file or envars defined for credentials
+    # this code makes me hate exceptions
+    try:
+        conn = get_s3_connection()
+    except S3ResponseError:
+        pytest.skip('unable to connect to s3')
+    else:
+        try:
+            grants = conn.get_bucket(test_bucket_name).get_acl().acl.grants
+        except S3ResponseError:
+            pytest.skip('no permission to read on bucket %s' %
+                        test_bucket_name)
+        else:
+            if not any(g.permission == 'FULL_CONTROL' or
+                       g.permission == 'READ' for g in grants):
+                pytest.skip('no permission to read on bucket %s' %
+                            test_bucket_name)
+            else:
+                return conn
+
+
+test_bucket_name = 'into-redshift-csvs'
+
+
+_tmps = ('tmp%d' % i for i in itertools.count())
+
+
+@pytest.yield_fixture
+def s3_bucket(conn):
+    b = 's3://%s/%s.csv' % (test_bucket_name, next(_tmps))
+    try:
+        yield b
+    finally:
+        drop(resource(b))
+
+
+@pytest.yield_fixture
+def s3_text_bucket(conn):
+    b = 's3://%s/%s.txt' % (test_bucket_name, next(_tmps))
+    try:
+        yield b
+    finally:
+        drop(resource(b))
+
+
+@pytest.yield_fixture
+def s3_json_bucket(conn):
+    b = 's3://%s/%s.json' % (test_bucket_name, next(_tmps))
+    try:
+        yield b
+    finally:
+        drop(resource(b))
+
+
+def test_s3_resource():
+    csv = resource(tips_uri)
+    assert isinstance(csv, S3(CSV))
+
+
+def test_s3_discover():
+    csv = resource(tips_uri)
+    assert isinstance(discover(csv), datashape.DataShape)
+
+
+def test_s3_to_local_csv():
+    with tmpfile('.csv') as fn:
+        csv = into(fn, tips_uri)
+        path = os.path.abspath(csv.path)
+        assert os.path.exists(path)
+
+
+def test_csv_to_s3_append(s3_bucket):
+    df = tm.makeMixedDataFrame()
+    s3 = resource(s3_bucket)
+    with tmpfile('.csv') as fn:
+        df.to_csv(fn, index=False)
+        append(s3, CSV(fn))
+    result = into(pd.DataFrame, s3)
+    tm.assert_frame_equal(df, result)
+
+
+def test_csv_to_s3_into(s3_bucket):
+    df = tm.makeMixedDataFrame()
+    with tmpfile('.csv') as fn:
+        df.to_csv(fn, index=False)
+        s3 = into(s3_bucket, CSV(fn))
+    result = into(pd.DataFrame, s3)
+    tm.assert_frame_equal(df, result)
+
+
+def test_s3_to_redshift(db):
+    redshift_uri = '%s::tips' % db
+    s3 = resource(tips_uri)
+    table = into(redshift_uri, s3, dshape=discover(s3))
+    dshape = discover(table)
+    ds = datashape.dshape("""
+    var * {
+        total_bill: ?float64,
+        tip: ?float64,
+        sex: ?string,
+        smoker: ?string,
+        day: ?string,
+        time: ?string,
+        size: int64
+    }""")
```

Comment:
maybe `assert discover(table) == discover(s3)` to avoid having to tweak this if we make changes to exactly what the datashapes should be?  Makes the test a little more *unit*y.